### DISCUSSION TITLE
Sponge layer application

### DISCUSSION THREAD

**Discussion by rafacmartins** [CliMA/Oceananigans.jl#3980] [Helpdesk]
Hi,
I’m currently running a few simulations of a submesoscale front confined in a zonal channel using Oceananigans. In my current setup, I need to apply a sponge layer at the y-boundaries of my domain as well as at the bottom (due to the propagation of internal waves, which can be observed around 14 days of simulation, 7-8s into the video below on the lower left panel, the plot it's not the best but it's just to illustrate those waves). 

https://github.com/user-attachments/assets/7ea8fb72-8938-49fb-8ac9-e223009d94d6

Note: this simulation above only applies the sponge layer only as `forcing = (u=mom_sponge, v=mom_sponge, w=mom_sponge)`, not applying something specific for those waves.

Currently, I’m planning to apply the sponge layer as follows, to make sure that those internal waves are not being reflected.

```
#--------------- Sponges

const sponge_size = 2.5kilometers
const slope = 1.25kilometers
x, y, z = nodes(grid, (Center(), Center(), Center()))
const ymin = minimum(y)
const ymax = maximum(y)

@inline mask_func(x,y,z) = ((
     tanh((y-(ymax-sponge_size))/slope)
    *tanh((y-(ymin+sponge_size))/slope)
)+1)/2

bottom_mask = GaussianMask{:z}(center=-grid.Lz, width=100meters)

mom_sponge = Relaxation(rate=1/30minutes, mask=mask_func, target=0)
uvw_sponge = Relaxation(rate=1/30minutes, mask=bottom_mask, target=0)

model = HydrostaticFreeSurfaceModel(; grid,
                                    coriolis = FPlane(latitude = 37),
                                    buoyancy = BuoyancyTracer(),
                                    tracers = :b,
                                    forcing = (u=mom_sponge, v=mom_sponge, w=bottom_mask ),
                                    momentum_advection = WENO(),
                                    tracer_advection = WENO())
#---------------------------
```

My question is: should I also apply the `mom_sponge` to w, or even apply the `bottom_mask` to u and v? If so, how could I do this? Would it look something like this?
`forcing = (u=(mom_sponge, bottom_mask), v=(mom_sponge, bottom_mask), w=(mom_sponge, bottom_mask) )`]

Thank you!

**Comment by simone-silvestri** [comment]
Hello @rafacmartins. 

Note that `w` is not evolved in the `HydrostaticFreeSurfaceModel`, but only diagnosed from the continuity equation. You cannot specify a forcing for `w` in the `HydrostaticFreeSurfaceModel`. i.e. the `w=bottom_mask` forcing in your script is doing nothing at the moment.  Therefore, applying the bottom_mask to u and v might be a good idea. A damping sponge for the horizontal velocities will typically also reduce vertical velocity due to continuity.

If you want to apply multiple forcings to a variable you indeed pass it as a tuple of forcings, so your syntax is in part correct (a part from the fact that `w` is not used).

However, you need to wrap the `bottom_mask` in a `Relaxation` forcing with a rate and a target (like you did for `uvw_sponge`). Therefore, you will end up with:
```julia
bottom_mask = GaussianMask{:z}(center=-grid.Lz, width=100meters)

north_sponge = Relaxation(rate=1/30minutes, mask=mask_func, target=0)
bottom_sponge = Relaxation(rate=1/30minutes, mask=bottom_mask, target=0)
mom_forcings =  (north_sponge, bottom_sponge)

forcing = (; u = mom_forcings, v = mom_forcings)
```

**Reply by rafacmartins** [reply]
Hello @simone-silvestri, thank you for the response!

It was my mistake that I didn't realize that even though we have `w` in the `HydrostaticFreeSurfaceModel` it's just an output from the continuity equation.

Also, thank you for the correction in the code, I was wrongly using `bottom_mask` instead of `uvw_sponge `. 

I applied the suggestion of using the `bottom_mask` in both `u` and `v` and it seems to work, at least in a lower resolution simulation, the reflection of those internal waves seems to have disappeared. I'm running one with the same resolution of the video above and expect to see a positive result.

Again, thank you for your help!

**Reply by rafacmartins** [reply]
Hi,

I stepped away from this issue for a while to focus on other parts of my work, but now I’m revisiting it. I’d like to ask if anyone has suggestions on how to verify whether the sponge layer at the bottom is effectively preventing the reflection of internal waves, beyond just relying on visual inspection. Any ideas or approaches would be greatly appreciated!

**Reply by glwagner** [reply]
The purpose of the sponge layer is basically to save computational costs, because otherwise you would need to make your domain deeper. So an appropriate test is to make your domain deeper, extending the sponge layer all the way to the bottom. An infinitely thick sponge is guaranteed to absorb internal waves, and eventually when the sponge becomes to thin it will stop working. If you look at the dependence of the part of the solution that you care about as a function of sponge layer thickness + domain depth, you should be able to deduce the point at which the sponge layer is "too thin".

**Reply by glwagner** [reply]
More generally as a basic strategy of thinking about tests --- think about your ultimate goals, like modeling some process. With any numerical settings, your objective is typically either minimize or eliminate the effect of the numerical settings on the conclusions you are coming to. So often, an efficient approach is to change the numerical settings (eg coarsening the resolution, changing the domain size if you have a sponge layer). Your goal is to find the range in which numerical settings _do_ affect the solution, and the range in which they don't, so that you can pick those settings to lie roughly between the two (ie, the most computationally efficient settings that do not pollute your result).

**Reply by rafacmartins** [reply]
Thanks for the help! I'll be more careful with these mistakes next time. In my case, the bottom sponge layer was correct, but I was placing it at the wrong center—putting it at -Lz when it should have been just 0.00. Now it's working perfectly. Thanks again!

**Reply by glwagner** [reply]
Great that you found it!
